### DISCUSSION TITLE
Lottie 4.0 more custom value providers for Core Animation engine

### DISCUSSION THREAD

**Discussion by opfeffer** [airbnb/lottie-ios#1836] [General]
Hi there- loved reading the announcement about the 4.0 release and the introduced performance improvements using `CoreAnimation`. I immediately went and updated our dependency from 3.4.3 to see the improvements in our app.

Unfortunately, I noticed that 4.0 still only supports the value provider for colors but no other types (see [ValueProviderStore.swift](https://github.com/airbnb/lottie-ios/blob/4.0.0/Sources/Private/CoreAnimation/ValueProviderStore.swift#L29-L32)).

I saw [this comment](https://github.com/airbnb/lottie-ios/discussions/1627#discussioncomment-3123069) in the 3.4 announcement discussion indicating that other custom value provider types were possibly coming in future releases. Is that still the plan? 

Congrats again to shipping a new major version of the library. Very excited to see its still being actively maintained! 👏

**Comment by calda** [comment]
Ideally this would be implemented eventually! I haven't prioritized it because I'm not really familiar with the use cases for this, other than customizing colors. Could you share information about your use case?

**Reply by opfeffer** [reply]
One trivial example is our use of `FloatValueProvider`s for `*.Opacity` transform layers to dynamically hide and show certain elements of animation. This [article](https://swiftsenpai.com/development/lottie-value-providers/) might be a good overview of other use cases such as  scale transforms..

**Reply by calda** [reply]
That opacity use case makes a ton of sense! Could you file an issue for this? I'll find a chance to add support for opacity here.

For now I probably plan on adding support for individual use cases as folks request them. That keeps things a bit more manageable, and will let us remove support for features that don't have much/any usage.

**Reply by opfeffer** [reply]
Done: https://github.com/airbnb/lottie-ios/issues/1837
